### DISCUSSION TITLE
Understanding the REST-API

### DISCUSSION THREAD

**Discussion by paderick** [dani-garcia/vaultwarden#4241] [Q&A]
Hello Vaultwarden-Community!
I want to understand how the API of vaultwarden (or bitwarden) works. After some research I found the OpenAPI documentation for [organization management](https://bitwarden.com/help/api/) and [vault management](https://bitwarden.com/help/vault-management-api/) from bitwarden directly. I guess that these will work as described, because the original clients (desktop / mobile) also work with vaultwarden as well.
Now i wanna use the (in `environment` object referred) values as described in [api/config](https://vaultwarden.self-hosted.com/api/config). For example the endpoint [/identity/connect/token](https://vaultwarden.self-hosted.com/identity/connect/token) for login or the endpoint [api/list/object/items](https://vaultwarden.self-hosted.com/api/list/object/items) for retrieving all items.

May someone explain me how I can access the REST-API from my self hosted vaultwarden instance please?

**Comment by BlackDex** [comment]
That is the documentation of the Public API, which Vaultwarden does not support (with a few exceptions).
Vaultwarden only supports the client API.

**Comment by paderick** [comment]
@BlackDex thank your for the fast response. Where can I find a documentation for the client API?

**Reply by zacknewman** [reply]
The Bitwarden source code:

* Bitwarden server: `git clone https://github.com/bitwarden/server`
* Bitwarden clients: `git clone https://github.com/bitwarden/clients`

I don't believe they have PDF, HTML, etc. documentation likely to allow them flexibility in changes. They make no guarantees on that API so it can change without notice.

You may also find the [`bitwarden-api-api`](https://crates.io/crates/bitwarden-api-api) crate useful.

**Comment by paderick** [comment]
The reason for my question was my intension to create native mobile apps (not as the xamarin app from bitwarden itself) to have a better experience on mobile devices. But to start a "vaultwarden goes mobile" initiative requires a good documentation beforehand.
I was thinking that there is somewhere a documentation from vaultwarden (because it seems like a community project).

**Comment by BlackDex** [comment]
Best would be to use the Bitwarden client repo as your source of truth.

As a side note, there already is a tool like that which also offers a premium version. It's called Keyguard.

**Reply by paderick** [reply]
I am asking for a (OpenAPI) documentation, because I am not familiar with the rust language. 😕

**Reply by BlackDex** [reply]
There currently is none.
Maybe this could be helpful https://github.com/bitwarden/sdk but I'm not sure if the client API is part of it to be fair,. didn't looked that deep into that repo.

**Comment by ubuntu811** [comment]
+1
Looking into deploying vaultwarden with a couple of organizations (teams) / users in a controlled environment. Teams can match already defined groups in ldap. I'll never win managing that manually, admin api would really help here.
